Bluetooth: controller: ISO-AL CIS RX framed and ISO bugfixes

mtpr-ot opened the pull request:
Adds reassembly of framed PDUs to ISO adaptation layer.
Reference times are also computed based on anchor points,
transmission latency components and a PDU specific time offset.

Some partial preparations for ISO broadcast can be found for
convenience. This needs to be properly merged, and currently will not
work with both CONFIG_BT_CTLR_CONN_ISO and CONFIG_BT_CTLR_SYNC_ISO
enabled.

Includes fixes and improvements for unframed mode.

Signed-off-by: Asger Munk Nielsen <asmk@oticon.com>
Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot commented:
The test `broadcast_iso.sh` fails. I assume this is due to some overlapping implementation. @cvinayak, could you please take a look at that?

cvinayak reviewed (changes requested):
Overlapping PR: https://github.com/zephyrproject-rtos/zephyr/pull/40602

Please have a separate PR for Broadcast ISO related file changes in this PR, so that I can take independent ISOAL changes for my testing. We can integrate broadcast ISO changes together thereafter.

mtpr-ot commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -591,6 +591,22 @@ config BT_CTLR_BROADCAST_ISO
 	bool
 	default BT_CTLR_ADV_ISO || BT_CTLR_SYNC_ISO
 
+config BT_CTLR_BROADCAST_ISO_STREAMS
+	int "LE Broadcast Isochronous Streams"
+	depends on BT_CTLR_BROADCAST_ISO
+	range 1 32
+	default 2
+	help
+	  Maximum supported total number of BISes.
+
```
I agree that having the ISO-AL and broadcast in separate PRs makes perfect sense. However, It is quite difficult to separate BIS/BIG out, as the code has now been designed to incorporate both, and logic just looks strange when removing one. Will try and see if it's possible.

cvinayak replied:
Thank you. There is only a little conflict with my other PR now.

mtpr-ot committed:
Bluetooth: controller: ISO-AL CIS RX framed and ISO bugfixes

Adds reassembly of framed PDUs to ISO adaptation layer.
Reference times are also computed based on anchor points,
transmission latency components and a PDU specific time offset.

Some partial preparations for ISO broadcast can be found for
convenience. This needs to be properly merged, and currently will not
work with both CONFIG_BT_CTLR_CONN_ISO and CONFIG_BT_CTLR_SYNC_ISO
enabled.

Includes fixes and improvements for unframed mode.

Signed-off-by: Asger Munk Nielsen <asmk@oticon.com>
Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -53,6 +53,8 @@
 #include "ll_sw/ull_sync_internal.h"
 #include "ll_sw/ull_conn_types.h"
 #include "ll_sw/ull_conn_internal.h"
+#include "ll_sw/isoal.h"
+#include "ll_sw/ull_iso_types.h"
```
I have conflict here. Minor, nothing that cannot be resolved.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_iso_types.h:
```diff
@@ -4,6 +4,33 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+/* CIS */
+#define LL_CIS_HANDLE_BASE CONFIG_BT_MAX_CONN
+
+#if defined(CONFIG_BT_CTLR_CONN_ISO_STREAMS)
+#define LAST_VALID_CIS_HANDLE \
+	(CONFIG_BT_CTLR_CONN_ISO_STREAMS + LL_CIS_HANDLE_BASE - 1)
+#else
+#define LAST_VALID_CIS_HANDLE (LL_CIS_HANDLE_BASE - 1)
+#endif /* CONFIG_BT_CTLR_CONN_ISO_STREAMS */
+
+#define LL_CIS_IDX_FROM_HANDLE(_handle) \
+	((_handle) - LL_CIS_HANDLE_BASE)
+
+#if defined(CONFIG_BT_CTLR_CONN_ISO)
+#define IS_CIS_HANDLE(_handle) \
+	(((_handle) >= LL_CIS_HANDLE_BASE) && \
+	((_handle) <= LAST_VALID_CIS_HANDLE))
+#else
+#define IS_CIS_HANDLE(_handle) 0
+#endif /* CONFIG_BT_CTLR_CONN_ISO */
+
+/* Common memebers for ll_conn_iso_stream and ll_broadcast_iso_stream */
+struct ll_iso_stream_hdr {
+	struct ll_iso_datapath *datapath_in;
+	struct ll_iso_datapath *datapath_out;
+};
+
```
This one we can refactor after this PR is merged.

cvinayak reviewed (commented):
Quick-ack, LGTM. Let me get this merged first, if its ok with you? https://github.com/zephyrproject-rtos/zephyr/pull/40602

mtpr-ot commented:
Rebased to include #40602. `ll_setup_iso_path` will need some refactoring to unite CIS- and BIS implementation. This PR will allow CIS **or** BIS, but not both configured at the same time.

mtpr-ot commented:
> Rebased to include #40602. `ll_setup_iso_path` will need some refactoring to unite CIS- and BIS implementation. This PR will allow CIS **or** BIS, but not both configured at the same time.

@cvinayak, iso_broadcast test fails - can you please take a look? Thanks.

cvinayak commented:
> @cvinayak, iso_broadcast test fails - can you please take a look? Thanks.

Sure. Working on it today.

cvinayak committed:
Bluetooth: Controller: Use a common ISO_PDU node rx type

Refactor implementation to use a single common ISO PDU
node rx type for ISO Synchronized Receiver.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot commented:
Looks good, @cvinayak. Thanks
